File: benchmark/10_spectral_norm/spectral_norm.go
// Benchmark 10: Spectral Norm
// Category: Linear Algebra
// Tests: Matrix-vector operations, power iteration

package main

import (
	"fmt"
	"math"
	"time"
)

const (
	N          = 1500
	ITERATIONS = 10
)

// A[i][j] = 1 / (i + j + 1)
func aElement(i, j int) float64 {
	return 1.0 / float64(i+j+1)
}

// Multiply by A
func av(v []float64, Av []float64) {
	for i := 0; i < N; i++ {
		Av[i] = 0.0
		for j := 0; j < N; j++ {
			Av[i] += aElement(i, j) * v[j]
		}
	}
}

// Multiply by A^T
func atv(v []float64, Atv []float64) {
	for i := 0; i < N; i++ {
		Atv[i] = 0.0
		for j := 0; j < N; j++ {
			Atv[i] += aElement(j, i) * v[j]
		}
	}
}

// Multiply by A^T * A
func atav(v, result []float64) {
	temp := make([]float64, N)
	av(v, temp)
	atv(temp, result)
}

func main() {
	start := time.Now()

	u := make([]float64, N)
	v := make([]float64, N)

	// Initialize vectors
	for i := 0; i < N; i++ {
		u[i] = 1.0
	}

	// Power iteration
	for iter := 0; iter < ITERATIONS; iter++ {
		atav(u, v)

		// Normalize
		norm := 0.0
		for i := 0; i < N; i++ {
			norm += v[i] * v[i]
		}
		norm = math.Sqrt(norm)

		for i := 0; i < N; i++ {
			u[i] = v[i] / norm
		}
	}

	// Calculate spectral norm approximation
	AvResult := make([]float64, N)
	av(u, AvResult)

	spectralNorm := 0.0
	for i := 0; i < N; i++ {
		spectralNorm += u[i] * AvResult[i]
	}
	spectralNorm = math.Sqrt(spectralNorm)

	elapsed := time.Since(start)

	fmt.Printf("Matrix size: %dx%d\n", N, N)
	fmt.Printf("Iterations: %d\n", ITERATIONS)
	fmt.Printf("Spectral norm: %.10f\n", spectralNorm)
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/10_spectral_norm/spectral_norm.rs
// Benchmark 10: Spectral Norm
// Category: Linear Algebra
// Tests: Matrix-vector operations, power iteration

use std::time::Instant;

const N: usize = 1500;
const ITERATIONS: usize = 10;

// A[i][j] = 1 / (i + j + 1)
fn a_element(i: usize, j: usize) -> f64 {
    1.0 / (i + j + 1) as f64
}

// Multiply by A
fn av(v: &[f64], av: &mut [f64]) {
    for i in 0..N {
        av[i] = (0..N).map(|j| a_element(i, j) * v[j]).sum();
    }
}

// Multiply by A^T
fn atv(v: &[f64], atv: &mut [f64]) {
    for i in 0..N {
        atv[i] = (0..N).map(|j| a_element(j, i) * v[j]).sum();
    }
}

// Multiply by A^T * A
fn atav(v: &[f64], result: &mut [f64]) {
    let mut temp = vec![0.0; N];
    av(v, &mut temp);
    atv(&temp, result);
}

fn main() {
    let start = Instant::now();

    let mut u = vec![1.0; N];
    let mut v = vec![0.0; N];

    // Power iteration
    for _ in 0..ITERATIONS {
        atav(&u, &mut v);

        // Normalize
        let norm: f64 = v.iter().map(|&x| x * x).sum::<f64>().sqrt();

        for i in 0..N {
            u[i] = v[i] / norm;
        }
    }

    // Calculate spectral norm approximation
    let mut av_result = vec![0.0; N];
    av(&u, &mut av_result);

    let spectral_norm: f64 = u.iter().zip(av_result.iter()).map(|(&a, &b)| a * b).sum();
    let spectral_norm = spectral_norm.sqrt();

    let elapsed = start.elapsed();

    println!("Matrix size: {}x{}", N, N);
    println!("Iterations: {}", ITERATIONS);
    println!("Spectral norm: {:.10}", spectral_norm);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
